include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -373,37 +374,31 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _Cp __combine, _
                     // 1. Initialization (transform part). Fill local memory
                     if (__is_first)
                     {
-                        __u(__item_id, __global_idx, __n, __temp_local, __rngs...);
+                        __u(__item_id, __n, __iters_per_witem, __temp_local, __rngs...);
                     }
                     else
                     {
-                        if (__global_idx < __n)
-                            __temp_local[__local_idx] = __temp_1_acc[__global_idx];
+                        if (__global_idx < __n_blocks)
+                            __temp_local[__local_idx] = __temp_1_acc[__offset_2 + __global_idx];
                         __item_id.barrier(sycl::access::fence_space::local_space);
                     }
                     // 2. Reduce within work group using local memory
-                    auto __res = __brick_reduce(__item_id, __global_idx, __n, __temp_local);
+                    auto __res = __brick_reduce(__item_id, __global_idx, __n_blocks, __temp_local);
                     if (__local_idx == 0)
                     {
-                        __temp_2_acc[__item_id.get_group(0)] = __res;
+                        __temp_1_acc[__offset_1 + __item_id.get_group(0)] = __res;
                     }
                 });
         });
         if (__is_first)
         {
-            __buf_2_ptr = &__temp_2;
             __is_first = false;
         }
-        else
-        {
-            ::std::swap(__buf_1_ptr, __buf_2_ptr);
-        }
-        __n = __n_groups;
-        __n_groups = (__n - 1) / __wgroup_size + 1;
-    } while (__n > 1);
-
-    //point of syncronization (on host access)
-    return __buf_1_ptr->template get_access<access_mode::read>()[0];
+        ::std::swap(__offset_1, __offset_2);
```

Comment:
Should we use fully qualified name? Or should we allow ADL work here?

I mean:

```cpp
using ::std::swap;
swap(off1, off2)
```

I understand that you don't change the behavior with your patch but probably the mistake has been made earlier.

Reply:
I don't see how ADL can help here. Could you please provide an example?

Reply:
It comes from unnecessary genericity I am talking about in the comment above.  Since I don't know the types of the offset I don't know what to expect. If it's just integral type, then you are right and everything should remain unchanged in that regard but genericity makes the code harder to read